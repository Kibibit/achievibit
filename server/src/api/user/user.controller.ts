import {
  Controller,
  Logger,
  NotFoundException,
  Param,
  UseFilters
} from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';

import { GetAll, GetOne } from '@kb-decorators';
import { KbValidationExceptionFilter } from '@kb-filters';
import { User } from '@kb-models';

import { UserService } from './user.service';

@Controller('api/user')
@ApiTags('user')
@UseFilters(new KbValidationExceptionFilter())
export class UserController {
  private readonly logger = new Logger(UserController.name);

  constructor(private readonly userService: UserService) {}

  @GetAll(User)
  async getAllUsers() {
    const users = await this.userService.findAllUsers();

    return users;
  }

  @GetOne(User, ':username')
  async getUser(@Param('username') username: string) {
    const user = await this.userService.findByUsername(username);

    if (!user) {
      throw new NotFoundException(`User with username ${ username } not found`);
    }

    // will show secret fields as well!
    this.logger.log('Full User');
    // will log only public fields!
    this.logger.log(user);
    // DANGER! WILL LOG EVERYTHING!
    // console.log(user);

    // will only include exposed fields
    return user;
  }
}
