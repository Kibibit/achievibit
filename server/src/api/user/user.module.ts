import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';

import { User } from '@kb-models';

import { UserController } from './user.controller';
import { UserService } from './user.service';

@Module({
  imports: [
    MongooseModule.forFeature([
      { name: User.modelName, schema: User.schema }
    ])
  ],
  providers: [UserService],
  controllers: [UserController]
})
export class UserModule {}
