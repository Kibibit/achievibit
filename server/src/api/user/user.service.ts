import { ReturnModelType } from '@typegoose/typegoose';

import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';

import { BaseService } from '@kb-abstracts';
import { User } from '@kb-models';

@Injectable()
export class UserService extends BaseService<User> {
  constructor(
    @InjectModel(User.modelName)
    private readonly userModel: ReturnModelType<typeof User>
  ) {
    super(userModel, User);
  }

  async findAllUsers(): Promise<User[]> {
    const dbUsers = await this.findAll().exec();

    return dbUsers.map((user) => new User(user.toObject()));
  }

  async findByUsername(username: string): Promise<User> {
    const dbUser = await this.findOne({ username }).exec();

    if (!dbUser) {
      return;
    }

    return new User(dbUser.toObject());
  }
}
