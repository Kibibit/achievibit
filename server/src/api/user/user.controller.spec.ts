import { noop } from 'lodash';

import { Test, TestingModule } from '@nestjs/testing';

import { DtoMockGenerator } from '@kb-dev-tools';

import { UserController } from './user.controller';
import { UserService } from './user.service';

describe.only('UserController', () => { 
  let controller: UserController;
  let userService: UserService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [UserController],
      providers: [{
        provide: UserService,
        useValue: { findByUsername: noop, findAllUsers: noop }
      }]
    }).compile();

    controller = module.get<UserController>(UserController);
    userService = module.get<UserService>(UserService);
  });

  it('should be defined', () => {
    expect(controller).toBeDefined();
  });

  it('should call service on get all users', async () => {
    const findAllResponse = DtoMockGenerator.users();

    // console.log('mock result: ', findAllResponse);

    const spyFindAll = jest.spyOn(userService, 'findAllUsers')
      .mockImplementation(() => Promise.resolve(findAllResponse));

    const result = await controller.getAllUsers();

    expect(result).toEqual(findAllResponse);
    expect(spyFindAll).toHaveBeenCalledTimes(1);
  });

  it('should get user by username', async () => {
    const findByUsernameResponse = DtoMockGenerator.user();

    // console.log('mock result: ', findAllResponse);
    const spyFindByUsername = jest.spyOn(userService, 'findByUsername')
      .mockImplementation(() => Promise.resolve(findByUsernameResponse));

    const result = await controller.getUser(findByUsernameResponse.username);

    expect(result).toEqual(findByUsernameResponse);
    expect(spyFindByUsername).toHaveBeenCalledTimes(1);
  });

  it('should throw error if user not found', async () => {
    const spyFindByUsername = jest.spyOn(userService, 'findByUsername')
      .mockImplementation(() => Promise.resolve(undefined));

    expect(controller.getUser('mockusername'))
      .rejects.toThrowErrorMatchingSnapshot();
    expect(spyFindByUsername).toHaveBeenCalledTimes(1);
  });
});
