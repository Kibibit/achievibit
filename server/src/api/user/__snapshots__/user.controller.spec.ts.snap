// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserController should throw error if user not found 1`] = `"User with username mockusername not found"`;
